katrain/gui/badukpan.py

```diff
@@ -253,91 +263,239 @@ def draw_board(self, *_args):
+                        or (self.gridpos[len(self.gridpos) - 1][len(self.gridpos[0]) - 1][0] in current_gridpos_x
+                            or self.gridpos[len(self.gridpos) - 1][len(self.gridpos[0]) - 1][1] in current_gridpos_y))
+            ):
+                self.resize_board()
+
+            self.draw_board_background(katrain, current_gridpos_x, current_gridpos_y, x_grid_spaces, y_grid_spaces, grid_spaces_margin_x, grid_spaces_margin_y)
+            self.draw_lines(current_gridpos_x, current_gridpos_y)
+            self.draw_star_points(board_size_x, board_size_y)
+            self.draw_coordinates(current_gridpos_x, current_gridpos_y)
+
+
+    def get_grid_spaces_margins(self):
+        if self.draw_coords_enabled:
+            grid_spaces_margin_x = [1.5, 0.75]  # left, right
+            grid_spaces_margin_y = [1.5, 0.75]  # bottom, top
+        else:  # no coordinates means remove the offset
+            grid_spaces_margin_x = [0.75, 0.75]  # left, right
+            grid_spaces_margin_y = [0.75, 0.75]  # bottom, top
+        return grid_spaces_margin_x, grid_spaces_margin_y
+
+
+    def calculate_grid_spaces(self, board_size_x, board_size_y, grid_spaces_margin_x, grid_spaces_margin_y):
+        x_grid_spaces = board_size_x - 1 + sum(grid_spaces_margin_x)
+        y_grid_spaces = board_size_y - 1 + sum(grid_spaces_margin_y)
+
+        return x_grid_spaces, y_grid_spaces
+
+
+    def calculate_grid_size(self, width, height, x_grid_spaces, y_grid_spaces):
+        # grid size is rounded to an integer to avoid rounding errors that
+        # produce tiny gaps between shaded grid squares
+        return math.floor(min(width / x_grid_spaces, height / y_grid_spaces) + 0.1)
+
+
+    def calculate_board_margins(self, x_grid_spaces, y_grid_spaces, grid_size):
+        board_width_with_margins = x_grid_spaces * grid_size
+        board_height_with_margins = y_grid_spaces * grid_size
+
+        return board_width_with_margins, board_height_with_margins
+
+
+    def calculate_extra_px_margins(self, width, height, board_width_with_margins, board_height_with_margins):
+        extra_px_margin_x = round((width - board_width_with_margins) / 2, 4)
+        extra_px_margin_y = round((height - board_height_with_margins) / 2, 4)
+
+        return extra_px_margin_x, extra_px_margin_y
+
+
+    def calculate_stone_size(self, grid_size):
+        return grid_size * Theme.STONE_SIZE
+
+
+    def initialize_gridpos(self):
+        self.gridpos = np.zeros((len(self.initial_gridpos_y), len(self.initial_gridpos_x)), dtype=object)
+        for y in range(len(self.initial_gridpos_y)):
+            for x in range(len(self.initial_gridpos_x)):
+                self.gridpos[y][x] = (self.initial_gridpos_x[x], self.initial_gridpos_y[y])
+
+
+    def initialize_gridpos_x_y(self, board_size_x, board_size_y, grid_spaces_margin_x, grid_spaces_margin_y, extra_px_margin_x, extra_px_margin_y, grid_size):
```

Comment:
combine with initialize_gridpos?

Reply:
I reuse initialize_gridpos_x_y in calculate_rotated_gridpos but not initialize_gridpos, so it makes sense to keep them separate I think